crates/holochain/src/core/state/cascade.rs

```diff
@@ -58,20 +55,106 @@ mod test;
+/////////////////
+
+/// Get an item from an option
+/// or return early from the function
+macro_rules! get_or_return {
+    ($n:expr) => {
+        match $n {
+            Some(n) => n,
+            None => return Ok(()),
+        }
+    };
+    ($n:expr, $ret:expr) => {
+        match $n {
+            Some(n) => n,
+            None => return Ok($ret),
+        }
+    };
+}
+
+/// Return from a function if
+/// an item is found otherwise continue
+macro_rules! return_if_found {
+    ($i:expr) => {
+        if let Some(i) = $i {
+            return Ok(Some(i));
+        }
+    };
+}
+
+/// Search every level that the cascade has been constructed with
+macro_rules! search_all {
+    ($cascade:expr, $fn:ident, $hash:expr) => {{
+        if let Some(db) = $cascade.pending_data.as_ref() {
+            return_if_found!($fn(db, $hash)?)
+        }
+        if let Some(db) = $cascade.judged_data.as_ref() {
+            return_if_found!($fn(db, $hash)?)
+        }
+        if let Some(db) = $cascade.integrated_data.as_ref() {
+            return_if_found!($fn(db, $hash)?)
+        }
+        if let Some(db) = $cascade.rejected_data.as_ref() {
+            return_if_found!($fn(db, $hash)?)
+        }
+        if let Some(db) = $cascade.cache_data.as_ref() {
+            let db = DbPair::from(db);
+            return_if_found!($fn(&db, $hash)?)
+        }
+        Ok(None)
+    }};
+}
+
+/// A pair containing an element buf and metadata buf
+/// with the same prefix.
+/// The default IntegratedPrefix is for databases that don't
+/// actually use prefixes (like the cache). In this case we just
+/// choose the first one (IntegratedPrefix)
+pub struct DbPair<'a, M, P = IntegratedPrefix>
 where
+    P: PrefixType,
+    M: MetadataBufT<P>,
+{
+    pub element: &'a ElementBuf<P>,
+    pub meta: &'a M,
+}
+
+pub struct DbPairMut<'a, M, P = IntegratedPrefix>
+where
+    P: PrefixType,
+    M: MetadataBufT<P>,
+{
+    pub element: &'a mut ElementBuf<P>,
+    pub meta: &'a mut M,
+}
+
+pub struct Cascade<
+    'a,
+    Network = HolochainP2pCell,
+    MetaVault = MetadataBuf,
+    MetaCache = MetadataBuf,
+    MetaPending = MetadataBuf<PendingPrefix>,
+    MetaJudged = MetadataBuf<JudgedPrefix>,
+    MetaRejected = MetadataBuf<RejectedPrefix>,
+> where
     Network: HolochainP2pCellT,
     MetaVault: MetadataBufT,
+    MetaPending: MetadataBufT<PendingPrefix>,
+    MetaJudged: MetadataBufT<JudgedPrefix>,
+    MetaRejected: MetadataBufT<RejectedPrefix>,
     MetaCache: MetadataBufT,
 {
-    element_vault: &'a ElementBuf,
-    meta_vault: &'a MetaVault,
-
-    element_cache: &'a mut ElementBuf,
-    meta_cache: &'a mut MetaCache,
-
-    env: EnvironmentRead,
-    network: Network,
+    integrated_data: Option<DbPair<'a, MetaVault, IntegratedPrefix>>,
+    pending_data: Option<DbPair<'a, MetaPending, PendingPrefix>>,
+    judged_data: Option<DbPair<'a, MetaJudged, JudgedPrefix>>,
+    rejected_data: Option<DbPair<'a, MetaRejected, RejectedPrefix>>,
```

Comment:
Isn't rejected data also judged data?

Reply:
OK I learned from you that Judged is post-validation, pre-integration, but Rejected is post-integration